sql: don't exclude unvalidated check constraints

thoszhang opened the pull request:
Previously, as of 19.1, AllActiveAndInactiveChecks(), which gets check
constraints for ALTER TABLE as well as SHOW CONSTRAINT, etc., excluded
unvalidated checks on the table descriptor, under the assumption that all
checks would be validated once they'd been added to the table descriptor. This
was to avoid double-counting constraints that were both on the table descriptor
and in the list of mutations. This caused unvalidated checks added in earlier
versions to disappear (though they are still on the table descriptor itself,
and are still enforced for writes). This PR fixes the issue by only excluding
`Validating` checks on the table descriptor to avoid double-counting.

Fixes #37285.

Release note (bug fix): Fixes bug where unvalidated check constraints disappear
from the output of SHOW CONSTRAINT and cannot be referenced in ALTER TABLE
after upgrading to 19.1.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/37366)
<!-- Reviewable:end -->

lucy-zhang committed:
sql: don't exclude unvalidated check constraints

Previously, as of 19.1, AllActiveAndInactiveChecks(), which gets check
constraints for ALTER TABLE as well as SHOW CONSTRAINT, etc., excluded
unvalidated checks on the table descriptor, under the assumption that all
checks would be validated once they'd been added to the table descriptor. This
was to avoid double-counting constraints that were both on the table descriptor
and in the list of mutations. This caused unvalidated checks added in earlier
versions to disappear (though they are still on the table descriptor itself,
and are still enforced for writes). This PR fixes the issue by only excluding
`Validating` checks on the table descriptor to avoid double-counting.

Release note (bug fix): Fixes bug where unvalidated check constraints disappear
from the output of SHOW CONSTRAINT and cannot be referenced in ALTER TABLE
after upgrading to 19.1.

lucy-zhang commented:
I updated the PR. I realized the previous approach (deduplicating by name, or even name and expression) doesn't work, since this function gets called in validating the uniqueness of constraint names before they're added.

What will work is just filtering out `Validating` constraints that are on the table descriptor to prevent them from being double-counted (and this should be a reasonable invariant to maintain even if we allow unvalidated constraints in the future, since unvalidated constraints can just be added to the descriptor and removed from the mutations list in a single version change).

lucy-zhang commented:
bors r+
